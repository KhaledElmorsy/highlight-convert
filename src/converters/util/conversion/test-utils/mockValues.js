export default [
  'usd',
  'egp',
  'eur',
  'gbp',
  'cad',
  'jpy',
  'aed',
].map((id, amount) => ({amount, unit: {id}}))
