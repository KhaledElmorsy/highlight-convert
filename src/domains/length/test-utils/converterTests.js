/**
 * Data set up 
 * {@link https://docs.google.com/spreadsheets/d/1Ml0Oc5__v8jOMlOIk5Wa-p6mJOhWEBeUZPC5KPf2AYk/edit?usp=sharing here}
 * (Google Docs)
 */
export default [
  [
    { ft: 6 },
    {
      m: 1.82871076,
      cm: 182.87107589,
      mm: 1828.71075891,
      km: 0.00182871,
      in: 71.99634258,
      ft: 6,
      yd: 1.99989637,
      mi: 0.00113631,
      thou: 71996.52544956,
    },
  ],
  [
    { cm: 200 },
    {
      m: 2,
      cm: 200,
      mm: 2000,
      km: 0.002,
      in: 78.74,
      ft: 6.562,
      yd: 2.1872,
      mi: 0.0012427,
      thou: 78740,
    },
  ],
  [
    { yd: 5 },
    {
      m: 4.572,
      cm: 457.2,
      mm: 4572,
      km: 0.004572,
      in: 180,
      ft: 15.001,
      yd: 5,
      mi: 0.0028409,
      thou: 180000,
    },
  ],
];
