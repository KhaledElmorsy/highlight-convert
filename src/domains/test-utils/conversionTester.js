export default function conversionTester(converter, unitMap) {
  return async (input, expected) => {
    const inputVector = Object.entries(input).reduce(
      (_, [id, amount]) => ({
        amount,
        unit: unitMap[id],
      }),
      null
    );

    const conversions = await converter.convert(inputVector);
    conversions.forEach(({ amount, unit }) => {
      expect(amount).toBe(expected[unit.id]);
    });
  };
}
