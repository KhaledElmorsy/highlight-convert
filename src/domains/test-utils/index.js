export { default as conversionTester } from './conversionTester';
export { default as matchTester } from './matchTester';
export { default as validateRenderSettings } from './validateRenderSettings';
