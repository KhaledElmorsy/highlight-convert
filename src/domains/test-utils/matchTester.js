/**
 * Test a unit matching method.
 * @param {(string: string) => Promise<Match[]>} match 
 * @param {{[id: Unit['id']]: Unit}} unitMap 
 * @returns 
 */
export default function matchTester(match, unitMap) {
  /**
   * @param {string} string
   * @param {{[id: Unit['id']]: number}[]} expected
   */
  return async (string, expectedValues) => {
    expect(await match(string)).toEqual(
      expect.arrayContaining(
        expectedValues.map((value) => {
          const [[id, amount]] = Object.entries(value);
          return expect.objectContaining({
            value: expect.objectContaining({
              unit: unitMap[id],
              amount,
            }),
          });
        })
      )
    );
  };
}
