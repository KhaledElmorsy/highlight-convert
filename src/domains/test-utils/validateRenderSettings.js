/**
 * Assert that all unit IDs in a render settings object belong to units in a
 * specified set.
 * @param {ConversionRenderSettings<Unit>} renderSettings
 * @param {{[id: string]: Unit}} unitMap
 */
export default function validateRenderSettings(renderSettings, unitMap) {
  const checkID = (id) => expect(Object.hasOwn(unitMap, id)).toBeTruthy();
  for (let [key, value] of Object.entries(renderSettings)) {
    switch (key) {
      case 'mainUnitID':
      case 'secondaryUnitID':
        checkID(value);
        break;
      case 'featuredUnitIDs':
        value.forEach(checkID);
        break;
      case 'groups':
        value.forEach((group) => {
          group.unitIDs.forEach(checkID);
        });
        break;
      case 'unitTemplates':
        Object.keys(value).forEach(checkID);
      // no-default
    }
  }
}
