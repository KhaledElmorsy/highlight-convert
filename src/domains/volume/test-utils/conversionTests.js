/**
 * Data set up with Google Docs
 * {@link https://docs.google.com/spreadsheets/d/1yDu63v41IcKgcL_Y0A442h9ttA9O4R7LfSMIXr5rrSU/edit?usp=sharing here}
 */
export default [
  [
    { l: 33.5 },
    {
      ml: 33500,
      l: 33.5,
      cm3: 33500,
      m3: 0.0335,
      tsp: 6796.6,
      tbsp: 2265.5,
      'fl oz': 1132.8,
      cup: 141.6,
      pt: 70.798,
      qt: 35.399,
      gal: 8.8498,
      in3: 2044.3,
      ft3: 1.183,
    },
  ],
  [
    { cup: 30 },
    {
      ml: 7097.6,
      l: 7.0976,
      cm3: 7097.6,
      m3: 0.0070976,
      tsp: 1440,
      tbsp: 480,
      'fl oz': 240,
      cup: 30,
      pt: 15,
      qt: 7.5,
      gal: 1.875,
      in3: 433.13,
      ft3: 0.25065,
    },
  ],
  [
    { 'fl oz': 0 },
    {
      ml: 0,
      l: 0,
      cm3: 0,
      m3: 0,
      tsp: 0,
      tbsp: 0,
      'fl oz': 0,
      cup: 0,
      pt: 0,
      qt: 0,
      gal: 0,
      in3: 0,
      ft3: 0,
    },
  ],
];
