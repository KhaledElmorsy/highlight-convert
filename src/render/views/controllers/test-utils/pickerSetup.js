// Prepped setup for views that pick from different passed options

const options = {
  string: ['testA', 'testB', 'testC'],
  object: [
    { id: 'a', name: 'Integer-test', type: 'number' },
    { id: 'b', name: 'Float-test', type: 'number' },
    { id: 'c', name: 'String-test', type: 'string' },
  ],
};

export const picker = {
  string: {
    options: options.string,
    mapOptions: (options) => options.toUpperCase(),
  },
  object: {
    options: options.object,
    mapOptions: ({ id, name, type }) => ({
      title: name.toUpperCase(),
      subtitle: id.toUpperCase(),
    }),
    keys: { main: 'title', sub: 'subtitle' },
  },
};

export const comboBox = {
  string: {
    options: options.string,
    mapOptions: (options) => options.toUpperCase(),
  },
  object: {
    options: options.object,
    mapOptions: ({ id, name, type }) => ({
      title: name.toUpperCase(),
      subtitle: id.toUpperCase(),
      group: type.toUpperCase(),
    }),
    keys: { main: 'title', sub: 'subtitle', group: 'group' },
  },
};
