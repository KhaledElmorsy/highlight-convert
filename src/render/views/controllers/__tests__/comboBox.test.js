import comboBox from '../combBox';
import { comboBox as setup } from '../test-utils/pickerSetup';

jest.mock('@ui5/webcomponents/dist/ComboBox', () => {});
jest.mock('@ui5/webcomponents/dist/MultiComboBox', () => {});

describe.each([
  ['Single', { multi: false, visitorMethod: 'comboBox' }],
  ['Multi', { multi: true, visitorMethod: 'multiComboBox' }],
])('%s option selection mode', (_, { multi, visitorMethod }) => {
  describe.each([
    [
      'String',
      {
        init: {
          options: setup.string.options,
          settings: { mapOptions: setup.string.mapOptions, multi },
        },
        values: multi
          ? {
              defaultValue: [setup.string.options[2]],
              newValue: setup.string.options.slice(0, 2),
            }
          : {
              defaultValue: setup.string.options[2],
              newValue: setup.string.options[0],
            },
        visitorArgs: {
          options: setup.string.options,
          keys: null,
          mapOptions: setup.string.mapOptions,
        },
      },
    ],
    [
      'Object',
      {
        init: {
          options: setup.object.options,
          settings: {
            mapOptions: setup.object.mapOptions,
            keys: setup.object.keys,
            multi,
          },
        },
        values: multi
          ? {
              defaultValue: [setup.object.options[2]],
              newValue: setup.object.options.slice(0, 2),
            }
          : {
              defaultValue: setup.object.options[2],
              newValue: setup.object.options[0],
            },
        visitorArgs: {
          options: setup.object.options,
          keys: setup.object.keys,
          mapOptions: setup.object.mapOptions,
        },
      },
    ],
  ])(
    '%s options',
    (_, { init, values: { defaultValue, newValue }, visitorArgs }) => {
      const testComboBox = comboBox({ ...init, value: defaultValue });

      it('Renders correctly', () => {
        expect(testComboBox).toMatchSnapshot();
      });
      it('Can programatically change values', () => {
        testComboBox.setValue(newValue);
        expect(testComboBox).toMatchSnapshot();
      });
      it('Dispatches the correct visitor method with relevant arguments', () => {
        const visitor = {
          comboBox: jest.fn(),
          multiComboBox: jest.fn(),
        };
        testComboBox.acceptVisitor(visitor);
        expect(visitor[visitorMethod]).toHaveBeenCalledWith({
          ...visitorArgs,
          element: testComboBox,
        });
      });
    }
  );
});
