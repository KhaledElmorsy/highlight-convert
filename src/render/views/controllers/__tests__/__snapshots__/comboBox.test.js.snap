// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Multi option selection mode Object options Can programatically change values 1`] = `
<ui5-multi-combobox>
  <ui5-mcb-group-item
    text="NUMBER"
  />
  <ui5-mcb-item
    additional-text="A"
    data-index="0"
    selected=""
    text="INTEGER-TEST"
  />
  <ui5-mcb-item
    additional-text="B"
    data-index="1"
    selected=""
    text="FLOAT-TEST"
  />
  <ui5-mcb-group-item
    text="STRING"
  />
  <ui5-mcb-item
    additional-text="C"
    data-index="2"
    text="STRING-TEST"
  />
</ui5-multi-combobox>
`;

exports[`Multi option selection mode Object options Renders correctly 1`] = `
<ui5-multi-combobox>
  <ui5-mcb-group-item
    text="NUMBER"
  />
  <ui5-mcb-item
    additional-text="A"
    data-index="0"
    text="INTEGER-TEST"
  />
  <ui5-mcb-item
    additional-text="B"
    data-index="1"
    text="FLOAT-TEST"
  />
  <ui5-mcb-group-item
    text="STRING"
  />
  <ui5-mcb-item
    additional-text="C"
    data-index="2"
    selected=""
    text="STRING-TEST"
  />
</ui5-multi-combobox>
`;

exports[`Multi option selection mode String options Can programatically change values 1`] = `
<ui5-multi-combobox>
  <ui5-mcb-item
    data-index="0"
    selected=""
    text="TESTA"
  />
  <ui5-mcb-item
    data-index="1"
    selected=""
    text="TESTB"
  />
  <ui5-mcb-item
    data-index="2"
    text="TESTC"
  />
</ui5-multi-combobox>
`;

exports[`Multi option selection mode String options Renders correctly 1`] = `
<ui5-multi-combobox>
  <ui5-mcb-item
    data-index="0"
    text="TESTA"
  />
  <ui5-mcb-item
    data-index="1"
    text="TESTB"
  />
  <ui5-mcb-item
    data-index="2"
    selected=""
    text="TESTC"
  />
</ui5-multi-combobox>
`;

exports[`Single option selection mode Object options Can programatically change values 1`] = `
<ui5-combobox
  value="INTEGER-TEST"
>
  <ui5-cb-group-item
    text="NUMBER"
  />
  <ui5-cb-item
    additional-text="A"
    data-index="0"
    text="INTEGER-TEST"
  />
  <ui5-cb-item
    additional-text="B"
    data-index="1"
    text="FLOAT-TEST"
  />
  <ui5-cb-group-item
    text="STRING"
  />
  <ui5-cb-item
    additional-text="C"
    data-index="2"
    text="STRING-TEST"
  />
</ui5-combobox>
`;

exports[`Single option selection mode Object options Renders correctly 1`] = `
<ui5-combobox
  value="STRING-TEST"
>
  <ui5-cb-group-item
    text="NUMBER"
  />
  <ui5-cb-item
    additional-text="A"
    data-index="0"
    text="INTEGER-TEST"
  />
  <ui5-cb-item
    additional-text="B"
    data-index="1"
    text="FLOAT-TEST"
  />
  <ui5-cb-group-item
    text="STRING"
  />
  <ui5-cb-item
    additional-text="C"
    data-index="2"
    text="STRING-TEST"
  />
</ui5-combobox>
`;

exports[`Single option selection mode String options Can programatically change values 1`] = `
<ui5-combobox
  value="TESTA"
>
  <ui5-cb-item
    data-index="0"
    text="TESTA"
  />
  <ui5-cb-item
    data-index="1"
    text="TESTB"
  />
  <ui5-cb-item
    data-index="2"
    text="TESTC"
  />
</ui5-combobox>
`;

exports[`Single option selection mode String options Renders correctly 1`] = `
<ui5-combobox
  value="TESTC"
>
  <ui5-cb-item
    data-index="0"
    text="TESTA"
  />
  <ui5-cb-item
    data-index="1"
    text="TESTB"
  />
  <ui5-cb-item
    data-index="2"
    text="TESTC"
  />
</ui5-combobox>
`;
